_posts/2024-12-02-safety-of-methods-for-numeric-primitive-types.md

```diff
@@ -0,0 +1,395 @@
+2. [Inexact Floating-Point Representations](https://github.com/model-checking/verify-rust-std/discussions/187): Not all decimal values can be exactly represented as floating-point numbers. When a value is assigned to a float, the actual stored value might slightly differ due to the limitations of floating-point precision.
+
+Unfortunately, despite multiple attempts, we were unable to derive an ideal solution. As a result, we submitted a [Kani feature request for `float_to_int_in_range`](https://github.com/model-checking/kani/issues/3711) method. The `float_to_int_in_range` method determines whether a float can be accurately represented in the target integer type, addressing the two challenges discussed above.
+
+### Last Try
+
+At this point, we had everything we needed. We updated the function contract to incorporate the `float_to_int_in_range` method.
+
+```rust
+// Preconditions
+// - Input float cannot be NaN
+// - Input float cannot be infinite
+// - Input float must be representable in the target integer type
+// `is_finite` handles the first two cases
+#[requires(self.is_finite() && float_to_int_in_range::<Self, Int>(self))]
+pub unsafe fn to_int_unchecked<Int>(self) -> Int where Self: FloatToInt<Int> { /* Implementation */ }
+```
+
+After running Kani again, all harnesses passed the verification checks. By this point, we completed the challenge.
+
+## Challenges Encountered & Lessons Learned
+
+Lastly, we summarized the main challenges we encountered throughout the course and our reflections.
+
+### Handling Exponential Input Spaces
+
+Verifying methods with large integer types was challenging due to the vast number of possible input combinations. To manage this, we implemented:
+
+- Partitioning Input Ranges: We targeted critical intervals where overflows are most likely, such as boundary values and mid-range points.
+- Using Assumptions: Leveraging `kani::assume`, we constrained inputs to these manageable ranges, ensuring comprehensive coverage of important cases without overwhelming the verifier.
+- Rationale: According to Randy E. Bryant's "Graph-Based Algorithms for Boolean Function Manipulation" (1986), the Boolean function representing all outputs of a multiplier as a BDD is exponential in the size of the operands. While a more compact representation can be achieved in CNF, it remains quadratic in the size of the operands. This complexity poses significant challenges for SAT solvers. Interestingly, this difficulty has practical implications: if solving such problems were more efficient, we could feasibly perform number factorization and thereby compromise some of today's cryptographic algorithms.
```

Comment:
Added this rationale behind the complexity involved in SAT solvers for handling larger input space.